tools/gfx/renderer-shared.cpp

```diff
@@ -96,4 +145,396 @@ Result createProgramFromSlang(IRenderer* renderer, IShaderProgram::Desc const& o
+    default:
+        break;
+    }
+    key.updateHash();
+    return getComponentId(key);
+}
+
+ShaderComponentID ShaderCache::getComponentId(UnownedStringSlice name)
+{
+    ComponentKey key;
+    key.typeName = name;
+    key.updateHash();
+    return getComponentId(key);
+}
+
+ShaderComponentID ShaderCache::getComponentId(ComponentKey key)
+{
+    ShaderComponentID componentId = 0;
+    if (componentIds.TryGetValue(key, componentId))
+        return componentId;
+    OwningComponentKey owningTypeKey;
+    owningTypeKey.hash = key.hash;
+    owningTypeKey.typeName = key.typeName;
+    owningTypeKey.specializationArgs.addRange(key.specializationArgs);
+    ShaderComponentID resultId = static_cast<ShaderComponentID>(componentIds.Count());
+    componentIds[owningTypeKey] = resultId;
+    return resultId;
+}
+
+void ShaderCache::init(ISlangFileSystem* cacheFileSystem)
+{
+    fileSystem = cacheFileSystem;
+
+    ComPtr<ISlangBlob> indexFileBlob;
+    if (fileSystem && fileSystem->loadFile("index", indexFileBlob.writeRef()) == SLANG_OK)
+    {
+        UnownedStringSlice indexText = UnownedStringSlice(static_cast<const char*>(indexFileBlob->getBufferPointer()));
+        TokenReader reader = TokenReader(indexText);
+        auto componentCountInFileSystem = reader.ReadUInt();
+        for (uint32_t i = 0; i < componentCountInFileSystem; i++)
+        {
+            OwningComponentKey key;
+            auto componentId = reader.ReadUInt();
+            key.typeName = reader.ReadWord();
+            key.specializationArgs.setCount(reader.ReadUInt());
+            for (auto& arg : key.specializationArgs)
+                arg = reader.ReadUInt();
+            componentIds[key] = componentId;
+        }
+    }
+}
+
+void ShaderCache::writeToFileSystem(ISlangMutableFileSystem* outputFileSystem)
+{
+    StringBuilder indexBuilder;
+    indexBuilder << componentIds.Count() << Slang::EndLine;
+    for (auto id : componentIds)
+    {
+        indexBuilder << id.Value << " ";
+        indexBuilder << id.Key.typeName << " " << id.Key.specializationArgs.getCount();
+        for (auto arg : id.Key.specializationArgs)
+            indexBuilder << " " << arg;
+        indexBuilder << Slang::EndLine;
+    }
+    outputFileSystem->saveFile("index", indexBuilder.getBuffer(), indexBuilder.getLength());
+    for (auto& binary : shaderBinaries)
+    {
+        ComPtr<ISlangBlob> blob;
+        binary.Value->writeToBlob(blob.writeRef());
+        outputFileSystem->saveFile(String(binary.Key).getBuffer(), blob->getBufferPointer(), blob->getBufferSize());
+    }
+}
+
+Slang::RefPtr<ShaderBinary> ShaderCache::tryLoadShaderBinary(ShaderComponentID componentId)
+{
+    Slang::ComPtr<ISlangBlob> entryBlob;
+    Slang::RefPtr<ShaderBinary> binary;
+    if (shaderBinaries.TryGetValue(componentId, binary))
+        return binary;
+
+    if (fileSystem && fileSystem->loadFile(String(componentId).getBuffer(), entryBlob.writeRef()) == SLANG_OK)
```

Comment:
Of course I *say* that the caching needs a lot of thought, and then you already have a working version of it implemented. :)

The main detail here is that the caching can only work because you've designed a keyspace that is entirely string-based. Maybe that is enough, but it is possible that sooner or later we will need a manifest/index format independent of the actual cached kernels.